ksm.go

```diff
@@ -0,0 +1,214 @@
+)
+
+var ksmSettings = map[string]ksmSetting{
+	ksmSlug:       {1000, 500}, // Every 500ms, we scan 1 page for every 1000 pages available in the system
+	ksmSlow:       {500, 100},  // Every 100ms, we scan 1 page for every 500 pages available in the system
+	ksmStandard:   {100, 10},   // Every 10ms, we scan 1 page for every 100 pages available in the system
+	ksmAggressive: {10, 1},     // Every ms, we scan 1 page for every 10 pages available in the system
+}
+
+const (
+	ksmRoot           = "/sys/kernel/mm/ksm/"
+	ksmRunFile        = "run"
+	ksmPagesToScan    = "pages_to_scan"
+	ksmSleepMillisec  = "sleep_millisecs"
+	ksmRun            = "1"
+	defaultKSMSetting = ksmAggressive
+)
+
+type sysfsAttribute struct {
+	path string
+	file *os.File
+}
+
+func (attr *sysfsAttribute) open() error {
+	file, err := os.OpenFile(attr.path, os.O_RDWR|syscall.O_NONBLOCK, 0660)
+	attr.file = file
+	return err
+}
+
+func (attr *sysfsAttribute) close() error {
+	err := attr.file.Close()
+	attr.file = nil
+
+	return err
+}
+
+func (attr *sysfsAttribute) read() (string, error) {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := ioutil.ReadAll(attr.file)
+	if err != nil {
+		return "", err
+	}
+
+	return string(data), nil
+}
+
+func (attr *sysfsAttribute) write(value string) error {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+	_, err = attr.file.WriteString(value)
+
+	return err
+}
+
+// store a sysfs attribute and returns its old value
+func (attr *sysfsAttribute) store(value string) (string, error) {
+	if attr.file == nil {
+		err := attr.open()
+		if err != nil {
+			return "", err
+		}
+
+		defer func() {
+			_ = attr.close()
+		}()
+	}
+
+	oldValue, err := attr.read()
+	if err != nil {
+		return "", err
+	}
+
+	err = attr.write(value)
+	if err != nil {
+		return "", err
+	}
+
+	return oldValue, nil
+}
+
+func isKSMAvailable(ksmDir string) error {
+	info, err := os.Stat(ksmDir)
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("%s is not available", ksmDir)
+	}
+
+	return nil
+}
+
+func pagesToScan(s ksmSetting) (string, error) {
+	return "10000", nil
+}
+
+var initialPagesToScan string
+var initialSleepInterval string
+var initialKSMRun string
+
+func ksmTune(s ksmSetting) error {
+	err := isKSMAvailable(ksmRoot)
+	if err != nil {
+		return err
+	}
+
+	pagesToScanSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmPagesToScan),
+	}
+
+	sleepIntervalSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmSleepMillisec),
+	}
+
+	runSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmRunFile),
+	}
+
+	newPagesToScan, err := pagesToScan(s)
+	if err != nil {
+		return err
+	}
+
+	initialPagesToScan, err = pagesToScanSysFs.store(newPagesToScan)
+	if err != nil {
+		return err
+	}
+
+	initialSleepInterval, err = sleepIntervalSysFs.store(fmt.Sprintf("%v", s.scanIntervalMs))
+	if err != nil {
+		fmt.Printf("Could not set slee interval %s\n", err)
```

Comment:
It's not clear to me if we want to use `Printf` here, or if we should be using the logger. proxy.go has a mix of logrus logging and print, but all the prints there go to os.stderr afaict. Can you clarify, confirm intention?

Reply:
Sorry, this should not be a printf, this is a debug leftover. Will fix.